docs(agents): Auto Agent Specification (#571)
gleanwork/glean-developer-site: docs/guides/agents/specification.mdx

> ---

Reviewer: 3/5 (strong opinion: non-blocking)

I think this guide is missing a link or description that explains what to do with one of these specifications (e.g. whether an api endpoint or some flow within the app).

It looks like it's meant as a target for some other "learn more" link; but can we add context around how to use a specification in this guide?

> title: Glean Auto Mode Agent Specification

Reviewer 2: Glean Agent Specification

description: A human-readable file system format for defining Glean Auto Mode agents.
---

# Glean Auto Mode Agent Specification

> The Glean Auto Mode agent specification is a human-readable file system format for defining a Glean Auto Mode agent — its instructions, tools, knowledge sources, skills, and subagents — as a directory of files. It is the source format consumed by the Glean Agents MCP tools and the GitHub Action, so agents can be authored and version-controlled outside the Glean UI.

Reviewer 2: nit: lets remove auto mode from all the places

> The Glean Auto Mode agent specification is a human-readable file system format for defining a Glean Auto Mode agent — its instructions, tools, knowledge sources, skills, and subagents — as a directory of files. It is the source format consumed by the Glean Agents MCP tools and the GitHub Action, so agents can be authored and version-controlled outside the Glean UI.

Reviewer 2: ```suggestion
The Glean Auto Mode agent specification is a human-readable file system format for defining a Glean Auto Mode agent — its instructions, tools, knowledge sources, skills, and subagents — as a directory of files. It is the source format consumed by the Glean Agents MCP tools and the GitHub Action, so Glean agents can be authored and version-controlled outside the UI.
```

## Directory layout

A typical agent directory looks like this:

```text
<agent-name>/
├── spec.yaml
├── instructions.md
├── skills/
│   └── <skill-name>/
│       └── SKILL.md
└── subagents/
    └── <subagent-name>/
        ├── spec.yaml
        ├── instructions.md
        └── skills/
```

Notes:

> - `spec.yaml` and `instructions.md` are required for the root agent.

Reviewer 2: ```suggestion
- `spec.yaml` and `instructions.md` are required for the root agent as well as for all the subagents.
```

- `skills/` is optional.
- `subagents/` is optional.

> - Skills do not have their own `spec.yaml`.

Reviewer 2: ```suggestion
- Skills should have `SKILL.md` file.
```

- Subagents can have skills, but cannot contain nested subagents.
- The directory names should use kebab-case.

> - The display name shown to users is defined in `spec.yaml`.

Reviewer 2: ```suggestion
- The display name of agents / subagents shown to users is defined in `spec.yaml`.
```

## Required files
### `spec.yaml`

`spec.yaml` is the main configuration file for the agent.

Common top-level fields:

| Field | Required | Description |
| --- | --- | --- |
| `id` | Yes | Stable unique agent identifier. Do not change it after creation. |
| `name` | Yes | Human-readable agent name shown to users. |
| `description` | Yes | Short description of what the agent does. |
| `instruction_file` | No | Path to the instructions file. Defaults to `instructions.md`. |
| `model` | No | Which agentic model the agent runs on and how hard it reasons. |

> | `trigger` | No | How the agent is invoked. |

Reviewer 2: ```suggestion
| `trigger` | No | How the agent is invoked. Defaults to chat based trigger. |
```

| `skills` | No | List of skill directory paths to include. |
| `subagents` | No | List of subagent directory paths to include. |
| `tools` | No | Tool providers and selected tools exposed to the agent. |
| `gleanSearchConfig` | No | Company-knowledge access configuration. |

> Example:

Reviewer 2: lets add model/trigger here as well.

```yaml
id: a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6
name: Travel Agent
description: >
  What this agent does, in a sentence or two.
instruction_file: instructions.md

skills:
  - skills/lead-qualification/
  - skills/deal-analysis/

subagents:
  - subagents/research-subagent/

tools:
  - toolProviderId: a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6
    selectedTools:
      - name: create-doc

>         customisationData:

Reviewer 2: customisationData is part of toolProvider now. so need to move one level up.

          skipUserInteraction: true
      - name: update-doc
      - name: search-doc
  - toolProviderId: 7e9f1a3b5c7d9e1f3a5b7c9d1e3f5a7b
    selectedTools:
      - name: send-email

gleanSearchConfig:
  datasourceInstances:
    - asana
    - confluence
  urls:

>     - https://drive.google.com/file/d/1F2MN8DfBTOEFxFzS3BA6OU7eQWpbEm3MLtpRtpDh0/view

Author: ```suggestion
    - https://drive.google.com/file/d/document/view
```

```

### `instructions.md`

`instructions.md` contains the agent's system prompt.

Guidelines:

- Use Markdown for structured workflows.
- Start with a clear role statement.
- Keep instructions precise and complete.
- Preserve exact step order when the workflow is rigid.
- Do not repeat tool descriptions unnecessarily.

## Optional configuration

### Model

The `model` block configures which agentic model the agent runs on and how hard it reasons. All keys are optional.

| Key | Description |
| --- | --- |
| `name` | Model Name |
| `mode` | Reasoning effort per turn. `ADVANCED` (default) or `FAST`. |

> | `autoUpgrade` | Boolean. When `true`, Glean rolls the agent onto newer model sets as they ship. |

Reviewer 2: ```suggestion
| `autoUpgrade` | Boolean. When `true`, Glean rolls the agent onto newer models as they ship. |
```

`mode` values:

- `ADVANCED` — high reasoning effort. The default when `mode` is omitted.

> - `FAST` — low reasoning effort.

Reviewer 2: ```suggestion
- `FAST` — low reasoning effort. This should be used only for simple tasks.
```

Example:

```yaml
model:
  name: GPT_5_4_MS
  mode: ADVANCED
  autoUpgrade: true
```

### Trigger

> `trigger` controls how the agent is invoked. Three modes are supported:

Reviewer 2: can we highlight the exact type values in supported trigger? like CHAT_MESSAGE / INPUT_FORM and also give sample examples.

- **Chat** — the user converses with the agent. This is the default when `trigger` is omitted.
- **Input form** — the agent receives a typed set of `trigger.inputFields` provided by the user at invocation time.
- **No input** — the agent runs without any user-supplied input; `trigger.inputFields` is empty.

### Tools

Tools are optional. Add them when the agent needs to perform actions outside its instructions and built-in knowledge, such as search, email, or API calls. If you declare tools, make sure every instructed external action has a corresponding tool.

Each tool entry includes:

> - `toolProviderId`: ID of the MCP server / Action Pack

Reviewer 2: ```suggestion
- `toolProviderId`: Glean internal identifier for tool provider which can also be MCP server.
```

> - `selectedTools` : list of actions

Reviewer 2: ```suggestion
- `selectedTools` : list of selected tools for the specified provider
```

> 

Reviewer 2: lets add customisationData details as well

### `gleanSearchConfig`

`gleanSearchConfig` controls company-knowledge access.

Supported fields:

| Field | Description |
| --- | --- |
| `datasourceInstances` | Restrict search to specific datasource instances. |
| `urls` | Restrict search to specific documents or folders. Folder URLs apply transitively to contained documents. |

Valid states:

| State | Meaning |
| --- | --- |
| Omitted | No company knowledge access. |
| `{}` | Unrestricted company knowledge access for the invoking user. |
| Populated | Restricted to the listed datasources and/or URLs. |

### Skills

A skill defines:

- what the skill does
- when to use it
- how it works

> Each skill directory must contain a `SKILL.md` file.

Reviewer 2: Display name on UI is directly derived from folder name

### Subagents

Subagents are optional isolated execution units.

Rules:

- A subagent has its own `spec.yaml` and `instructions.md`.
- A subagent may have its own `skills/` directory.

> - A subagent must not declare its own `subagents` field.

Reviewer 2: ```suggestion
- A subagent must not declare its own `subagents` field since nested subagents are not supported
```

remove next line

---
title: Glean Agent Specification
description: A human-readable file system format for defining Glean agents.
---

# Glean Agent Specification

The Glean agent specification is a human-readable file system format for defining a Glean agent — its instructions, tools, knowledge sources, skills, and subagents — as a directory of files. It is the source format consumed by the Glean Agents MCP tools and the GitHub Action, so Glean agents can be authored and version-controlled outside the UI.

> Once authored, pass the specification directory to the Glean Agents MCP tools or the GitHub Action to create or update agents in Glean.

Reviewer: 3/5 (strong opinion: non-blocking)

Can we add useful links (e.g. to docs) to
"Glean Agents MCP tools" and
"GitHub Action"?

Author: Yes, next in plan.

### Model

The `model` block configures which agentic model the agent runs on and how hard it reasons. All keys are optional.

| Key | Description |
| --- | --- |
| `name` | Model Name |
| `mode` | Reasoning effort per turn. `ADVANCED` (default) or `FAST`. |
| `autoUpgrade` | Boolean. When `true`, Glean rolls the agent onto newer models as they ship. |

`mode` values:

- `ADVANCED` — high reasoning effort. The default when `mode` is omitted.
- `FAST` — low reasoning effort. This should be used only for simple tasks.

Example:

```yaml
model:
  name: GPT_5_4_MS
  mode: ADVANCED
  autoUpgrade: true
```

### Trigger

`trigger` controls how the agent is invoked.

Supported `type` values:

| `type` | Description |
| --- | --- |
| `CHAT_MESSAGE` | The user converses with the agent. This is the default when `trigger` is omitted. |
| `INPUT_FORM` | The agent receives a typed set of `trigger.inputFields` provided by the user at invocation time. When `inputFields` is empty, the agent runs without any user input. |

Examples:

```yaml
# Chat trigger (default)
trigger:
  type: CHAT_MESSAGE
```

```yaml
# Input form trigger
trigger:
  type: INPUT_FORM
  inputFields:
    - name: query

>       label: Search query

Reviewer 2: should add default and type as well. and specify all supported types
